src/oscar/apps/offer/abstract_models.py

```diff
@@ -443,6 +451,13 @@ def products(self):
         return queryset.filter(is_discountable=True).exclude(
             structure=Product.CHILD)
 
+    @cached_property
+    def combined_offers(self):
```

Comment:
I think this can just be done in a single query.

Something like:
```
ConditionalOffer.filter(
    models.Q(id__in=self.pk)
    |models.Q(id__in=self.combinations.values_list("id", flat=True)
    |models.Q(id__in=self.in_combination.values_list("id", flat=True)
)
```
that way you can do more filtering on the resulting queryset.